outlines/fsm/parsing.py

```diff
@@ -626,9 +666,22 @@ def lex(self, lexer_state: LexerState, parser_state: Any) -> Iterator[Token]:
         try:
             while True:
                 lexer = self.lexers[parser_state.position]
-                yield lexer.next_token(lexer_state, parser_state)
+                next_tok = lexer.next_token(lexer_state, parser_state)
+                yield next_tok
         except EOFError:
             pass
+        except KeyError:
+            if len(lexer_state.text) > lexer_state.line_ctr.char_pos:
+                raise UnexpectedCharacters(
+                    lexer_state.text,
+                    lexer_state.line_ctr.char_pos,
+                    lexer_state.line_ctr.line,
+                    lexer_state.line_ctr.column,
+                    allowed=False,
+                    token_history=lexer_state.last_token and [lexer_state.last_token],
+                    state=parser_state,
+                    terminals_by_name=self.root_lexer.terminals,
+                )
```

Comment:
Note: Fixes the following tests in `test_cfg_guide.py::test_cfg_next_token`

- Multiple Valid Continuations
- Token is Substring of Another Token
- Recursive Patterns